source/_posts/2017-11-15-code-generation-with-sourcery.md

```diff
@@ -0,0 +1,133 @@
+        static let companyId = "companyId"
+    }
+    
+    public let storage = Storage()
+    
+    public var email: String
+    public var name: String?
+    public var password: String
+    public var companyId: Identifier
+    
+    public init(
+        email: String,
+        name: String?,
+        password: String,
+        companyId: Identifier
+    ) {
+        self.email = email
+        self.name = name
+        self.password = password
+        self.companyId = companyId
+    }
+
+    public init(row: Row) throws {
+        name = try row.get(Keys.name)
+        email = try row.get(Keys.email)
+        password = try row.get(Keys.password)
+        companyId = try row.get(Keys.companyId)
+    }
+  
+    internal enum JSONKeys: String {
+        case name
+        case email
+        case companyId
+    }
+}
+
+extension AppUser: JSONRepresentable {
+    public func makeJSON() throws -> JSON {
+        var json = JSON()
+        try json.set(idKey, id)
+        try json.set(JSONKeys.name.rawValue, name)
+        try json.set(JSONKeys.email.rawValue, email)
+        try json.set(JSONKeys.companyId.rawValue, companyId)
+        return json
+    }
+}
+
+extension AppUser: NodeRepresentable {
+    public func makeNode(in context: Context?) throws -> Node {
+        var node = Node([:])
+        try node.set(idKey, id)
+        try node.set(Keys.name, name)
+        try node.set(Keys.email, email)
+        try node.set(Keys.companyId, companyId)
+        try node.set(AppUser.createdAtKey, createdAt)
+        try node.set(AppUser.updatedAtKey, updatedAt)
+        try node.set("company", company.get())
+        return node
+    }
+}
+
+extension AppUser: Preparation {
+    public static func prepare(_ database: Database) throws {
+        try database.create(self) { user in
+            user.id()
+            user.string(Keys.email, unique: true)
+            user.string(Keys.name, optional: true)
+            user.string(Keys.password)
+            user.string(Keys.companyId)
+        }
+        
+        
+        try database.index("email", for: AppUser.self)
+        try database.index(Keys.companyId, for: AppUser.self)
+    }
+    
+    public static func revert(_ database: Database) throws {
+        try database.delete(self)
+    }
+}
+
+extension AppUser {
+    public func makeRow() throws -> Row {
+        var row = Row()
+        try row.set(Keys.email, email)
+        try row.set(Keys.name, name)
+        try row.set(Keys.password, password)
+        try row.set(Keys.companyId, companyId)
+        return row
+    }
+}
+```
+
+As you can see, it's quite a bit of code. Sadly, it doesn't get much better from here. The more Vapor features you opt-in to the more boilerplate you're going to have to write. At Nodes, we noticed that most of our day was spent writing boilerplate and we began to search for solutions. After doing some research, we ended up going with [Sourcery](https://github.com/krzysztofzablocki/Sourcery).
```

Comment:
It might be implicit from the word "boilerplate", but I think it makes sense to quickly describe that the code in the example above doesn't change a lot from model to model and is generally "simple". That is why we are able to automate this by using metaprogramming. Maybe also mention that there is nothing in the Vapor toolbox that does this (or scaffolding)